lib/train/transports/cisco_ios.rb

```diff
@@ -0,0 +1,147 @@
+# encoding: utf-8
+
+require 'train/plugins'
+require 'train/transports/ssh'
+
+module Train::Transports
+  class BadEnablePassword < Train::TransportError; end
+
+  class CiscoIOS < SSH
+    name 'cisco_ios'
+
+    option :host, required: true
+    option :user, required: true
+    option :port, default: 22, required: true
+
+    option :password, required: true
+
+    # Used to elevate to enable mode (similar to `sudo su` in Linux)
+    option :enable_password
+
+    def connection
+      validate_options(@options)
+      @connection ||= Connection.new(@options)
+    end
+
+    class Connection < BaseConnection
+      def initialize(options)
+        super(options)
+
+        @host = @options.delete(:host)
+        @user = @options.delete(:user)
+        @port = @options.delete(:port)
+
+        @enable_password = @options.delete(:enable_password)
+
+        @prompt = /^\S+[>#]\r\n.*$/
+      end
+
+      def uri
+        "ssh://#{@user}@#{@host}:#{@port}"
+      end
+
+      private
+
+      def establish_connection
+        logger.debug("[SSH] opening connection to #{self}")
+
+        @ssh = Net::SSH.start(
+          @host,
+          @user,
+          @options.delete_if { |_key, value| value.nil? },
+        )
+
+        @channel ||= open_channel
+
+        # Escalate privilege to enable mode if password is given
+        if @enable_password
+          run_command_via_channel("enable\r\n#{@enable_password}")
+        end
+
+        # Prevent `--MORE--` by removing terminal length limit
+        run_command_via_channel('terminal length 0')
+
+        @ssh
+      end
+
+      def run_command_via_connection(cmd)
+        @session ||= establish_connection
+
+        result = run_command_via_channel(cmd)
+        CommandResult.new(*format_result(result))
+      end
+
+      def format_result(result)
+        stderr_with_exit_1 = ['', result, 1]
+        stdout_with_exit_0 = [result, '', 0]
+
+        # IOS commands do not have an exit code, so we must capture known errors
+        case result
+        when /Bad IP address/
+          stderr_with_exit_1
+        when /Incomplete command/
+          stderr_with_exit_1
+        when /Invalid input detected/
+          stderr_with_exit_1
+        when /Unrecognized host/
+          stderr_with_exit_1
+        else
+          stdout_with_exit_0
+        end
```

Comment:
As far as better ways to organize this, it seems like they're all just string literals (`when 'Bad IP address'` and so forth would have worked too) and they all should be treated as exit 1, so this is what came to mind. Basically just extract the data from the logical flow.

```ruby
ERROR_STRINGS = [
  'Bad IP address',
  'Incomplete command',
  'Invalid input detected',
  'Unrecognized host'
]

# IOS commands do not have an exit code, so we must capture known errors
def format_result(result)
  stderr_with_exit_1 = ['', result, 1]
  stdout_with_exit_0 = [result, '', 0]

  ERROR_strings.include?(result) ? stderr_with_exit_1 : stdout_with_exit_0
end
```
I debated inlining the return values, but I really like those explaining variables!